In [1]:
import pandas as pd
import numpy as np
from scipy.io import loadmat

## Task 1

The data is already split for us so we wont need to use the train_test_split function but we'll need to seperate features and labels

In [2]:
# Loading data from .mat files using the loadmat function
train_data = loadmat("./task 1 data/train.mat")
val_data = loadmat("./task 1 data/val.mat")
test_data = loadmat("./task 1 data/test.mat")

In [3]:
# Seperating features and labels
train_X = train_data['features']
train_Y = train_data["label"].flatten()

val_X = val_data['features']
val_Y = val_data["label"].flatten()

test_X = test_data['features']
test_Y = test_data["label"].flatten()

## Decision Tree

In [4]:
from sklearn.tree import DecisionTreeClassifier

dtree = DecisionTreeClassifier(max_depth=15, random_state=43)
dtree = dtree.fit(train_X, train_Y)

#### Validation

In [5]:
from sklearn.metrics import accuracy_score
pred = dtree.predict(val_X)
print(f'Validation accuracy: {accuracy_score(val_Y, pred)}')

Validation accuracy: 0.825


#### Testing

In [6]:
pred = dtree.predict(test_X)
print(f'Test accuracy: {accuracy_score(test_Y, pred)}')

Test accuracy: 0.8138888888888889


#### Conclusion
The Highest Accuracy I'm able to get is 82.5% on validation and 81.4% on test which is not great but not bad. The max depth of 15 yielded this accuracy. A higher number didn't change the accuracy while a lower depth caused a my accuracy to decrease so I stuck with 15. 

However we need an accuracy of 95% so we'll check other models.

# ------------------------------------------------------------------------------------------------

## K-NN 

In [7]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(1)
knn = knn.fit(train_X, train_Y)

#### Validation

In [8]:
pred = knn.predict(val_X)
print(f'Validation accuracy: {accuracy_score(val_Y, pred)}')

Validation accuracy: 0.9888888888888889


#### Testing

In [9]:
pred = knn.predict(test_X)
print(f'Test accuracy: {accuracy_score(test_Y, pred)}')

Test accuracy: 0.9777777777777777


#### Conclusion
The Highest Accuracy I'm able to get is 98.9% on validation and 97.8% on test which is extremely good. I found that increasing the number of neighbors decreased my accuracy, though not by alot, so 1 neighbor worked perfectly with no overfitting for this data however overfitting might be seen when a larger data set is tested so it might be wise to increase the number of neighbors a little.

# ------------------------------------------------------------------------------------------------

## Logistic Regression

In [10]:
from sklearn.linear_model import LogisticRegression

reg = LogisticRegression(solver='lbfgs', max_iter=1500, random_state=43)
reg = reg.fit(train_X, train_Y)

#### Validation

In [11]:
pred = reg.predict(val_X)
print(f'Validation accuracy: {accuracy_score(val_Y, pred)}')

Validation accuracy: 0.9666666666666667


#### Testing

In [12]:
pred = reg.predict(test_X)
print(f'Validation accuracy: {accuracy_score(test_Y, pred)}')

Validation accuracy: 0.9638888888888889


#### Conclusion
The Highest Accuracy I'm able to get is 96.7% on validation and 96.4% on test which is also extremely good. Using a different solver such as 'liblinear' allowed me to decrease the max_iter parameter however it also decreased the accuracy so I used a higher max_iter with the defualt solver, 'lbfgs'.

# ------------------------------------------------------------------------------------------------

# Best to Worst (95% accuracy threshold)
1. KNN
2. Logistic Regression
3. Decision Tree